src/Nethermind/Nethermind.Core/Crypto/KeccakCache.cs

```diff
@@ -0,0 +1,207 @@
+        Memory = (Entry*)NativeMemory.AlignedAlloc(size, Entry.Size);
+        NativeMemory.Clear(Memory, size);
+    }
+
+    [SkipLocalsInit]
+    public static ValueHash256 Compute(ReadOnlySpan<byte> input)
+    {
+        Unsafe.SkipInit(out ValueHash256 hash);
+
+        // Special cases first
+        if (input.Length == 0)
+        {
+            hash = ValueKeccak.OfAnEmptyString;
+            goto Return;
+        }
+
+        if (input.Length > Entry.MaxPayloadLength)
+        {
+            hash = ValueKeccak.Compute(input);
+            goto Return;
+        }
+
+        var fast = FastHash(input);
+        var index = fast & BucketMask;
+
+        Debug.Assert(index < Count);
+
+        uint hashAndLength = (fast & HashMask) | (ushort)input.Length;
+
+        ref Entry e = ref Unsafe.Add(ref Unsafe.AsRef<Entry>(Memory), index);
+
+        // Read aligned, volatile, won't be torn, check with computed
+        if (Volatile.Read(ref e.HashAndLength) == hashAndLength)
+        {
+            // There's a possibility of a hit, try lock.
+            if (Interlocked.CompareExchange(ref e.Lock, Entry.Locked, Entry.Unlocked) == Entry.Unlocked)
+            {
+                if (e.HashAndLength != hashAndLength)
+                {
+                    // The value has been changed between reading and taking a lock.
+                    // Release the lock and compute.
+                    Volatile.Write(ref e.Lock, Entry.Unlocked);
+                    goto Compute;
+                }
+
+                // Local copy of 128 bytes, to release the lock as soon as possible and make a key comparison without holding it.
+                Entry copy = e;
+
+                // Release the lock
+                Volatile.Write(ref e.Lock, Entry.Unlocked);
+
+                // Lengths are equal, the input length can be used without any additional operation.
+                if (MemoryMarshal.CreateReadOnlySpan(ref copy.Payload, input.Length).SequenceEqual(input))
+                {
+                    hash = copy.Value;
+                    goto Return;
+                }
+            }
+        }
+
+    Compute:
+        hash = ValueKeccak.Compute(input);
+
+        // Try lock and memoize
+        if (Interlocked.CompareExchange(ref e.Lock, Entry.Locked, Entry.Unlocked) == Entry.Unlocked)
+        {
+            e.HashAndLength = hashAndLength;
+            e.Value = hash;
+
+            input.CopyTo(MemoryMarshal.CreateSpan(ref e.Payload, input.Length));
+
+            // Release the lock
+            Volatile.Write(ref e.Lock, Entry.Unlocked);
+        }
+
+    Return:
+        return hash;
+    }
+
+    /// <summary>
+    /// Gets the bucket for tests.
+    /// </summary>
+    public static uint GetBucket(ReadOnlySpan<byte> input) => FastHash(input) & BucketMask;
+
+    [SkipLocalsInit]
+    [MethodImpl(MethodImplOptions.AggressiveInlining)]
+    private static uint FastHash(ReadOnlySpan<byte> input)
```

Comment:
Hash from end rather than start. The end of the bytes is higher entropy than the start

Reply:
nvm, you use the whole length